src/mito2/src/memtable/time_partition.rs

```diff
@@ -0,0 +1,543 @@
+
+    /// Freezes all memtables.
+    pub fn freeze(&self) -> Result<()> {
+        let inner = self.inner.lock().unwrap();
+        for part in &*inner.parts {
+            part.memtable.freeze()?;
+        }
+        Ok(())
+    }
+
+    /// Forks latest partition.
+    pub fn fork(&self, metadata: &RegionMetadataRef) -> Self {
+        let mut inner = self.inner.lock().unwrap();
+        let latest_part = inner
+            .parts
+            .iter()
+            .max_by_key(|part| part.time_range.map(|range| range.min_timestamp))
+            .cloned();
+
+        let Some(old_part) = latest_part else {
+            return Self::new(
+                metadata.clone(),
+                self.builder.clone(),
+                inner.next_memtable_id,
+                self.part_duration,
+            );
+        };
+        let memtable = old_part.memtable.fork(inner.alloc_memtable_id(), metadata);
+        let new_part = TimePartition {
+            memtable,
+            time_range: old_part.time_range,
+        };
+        Self {
+            inner: Mutex::new(PartitionsInner::with_partition(
+                new_part,
+                inner.next_memtable_id,
+            )),
+            part_duration: self.part_duration,
+            metadata: metadata.clone(),
+            builder: self.builder.clone(),
+        }
+    }
+
+    /// Returns partition duration.
+    pub(crate) fn part_duration(&self) -> Option<Duration> {
+        self.part_duration
+    }
+
+    /// Returns memory usage.
+    pub(crate) fn memory_usage(&self) -> usize {
+        let inner = self.inner.lock().unwrap();
+        inner
+            .parts
+            .iter()
+            .map(|part| part.memtable.stats().estimated_bytes)
+            .sum()
+    }
+
+    /// Append memtables in partitions to small vec.
+    pub(crate) fn list_memtables_to_small_vec(&self, memtables: &mut SmallMemtableVec) {
+        let inner = self.inner.lock().unwrap();
+        memtables.extend(inner.parts.iter().map(|part| part.memtable.clone()));
+    }
+
+    /// Returns the next memtable id.
+    pub(crate) fn next_memtable_id(&self) -> MemtableId {
+        let inner = self.inner.lock().unwrap();
+        inner.next_memtable_id
+    }
+
+    /// Returns all partitions.
+    fn list_partitions(&self) -> PartitionVec {
+        let inner = self.inner.lock().unwrap();
+        inner.parts.clone()
+    }
+
+    /// Write to multiple partitions.
+    fn write_multi_parts(&self, kvs: &KeyValues, parts: &PartitionVec) -> Result<()> {
+        let mut parts_to_write = HashMap::new();
+        let mut missing_parts = HashMap::new();
+        for kv in kvs.iter() {
+            let mut part_found = false;
+            // Safety: We used the timestamp before.
+            let ts = kv.timestamp().as_timestamp().unwrap().unwrap();
+            for part in parts {
+                if part.contains_timestamp(ts) {
+                    // Safety: If time range is None then we won't go to this method.
```

Comment:
But in `contains_timestamp` method, if `time_range` is None, it returns true: 
```rust
let Some(range) = self.time_range else {
    return true;
};
```

Reply:
There is an invariant: If part_duration is None, then there will be only one partition and the time range of the partition is None; If part_duration is Some, then all partitions' time ranges are always Some.